libraries/abstractions/pkcs11/mbedtls/iot_pkcs11_mbedtls.c

```diff
@@ -219,43 +222,74 @@ typedef struct P11Session
  */
-#define PKCS11_MODULE_IS_INITIALIZED    ( ( xP11Context.xIsInitialized == CK_TRUE ) ? CK_TRUE : CK_FALSE )                                                                                          /**< Checks if PKCS #11 module is initialized. */
-#define PKCS11_SESSION_IS_OPEN( xSessionHandle )                         ( ( ( ( ( P11SessionPtr_t ) xSessionHandle )->xOpened ) == CK_TRUE ) ? CKR_OK : CKR_SESSION_CLOSED )                       /**< Checks if the current session is open */
-#define PKCS11_SESSION_IS_VALID( xSessionHandle )                        ( ( ( P11SessionPtr_t ) xSessionHandle != NULL ) ? PKCS11_SESSION_IS_OPEN( xSessionHandle ) : CKR_SESSION_HANDLE_INVALID ) /**< Checks if the current session is valid */
-#define PKCS11_SESSION_VALID_AND_MODULE_INITIALIZED( xSessionHandle )    ( PKCS11_MODULE_IS_INITIALIZED ? PKCS11_SESSION_IS_VALID( xSessionHandle ) : CKR_CRYPTOKI_NOT_INITIALIZED )                /**< Checks if the current session is valid and initialized. */
-/*-----------------------------------------------------------*/
-/*--------- See iot_pkcs11_pal.c for definitions ------------*/
+static CK_RV prvCheckValidSessionAndModule( const P11Session_t * pxSession )
+{
+    CK_RV xResult = CKR_OK;
 
+    /* An integer constant expression with the value 0 or 1 of either signedness
+     * may be cast to a type which is defined as essentially Boolean.
+     * This allows the implementation of non-C99 Boolean models.
+     * The PKCS #11 standard defines the boolean type as CK_BBOOL, whose
+     * underlying type is an unsigned char, therefore conforming to the MISRA
+     * exception.
+     */
+    /* coverity[misra_c_2012_rule_10_5_violation] */
+    if( xP11Context.xIsInitialized == ( CK_BBOOL ) CK_FALSE )
+    {
+        xResult = CKR_CRYPTOKI_NOT_INITIALIZED;
+    }
+    else if( pxSession == NULL )
+    {
+        xResult = CKR_SESSION_HANDLE_INVALID;
+    }
+    else
+    {
+        /* Session is initialized and valid. */
+    }
 
+    return xResult;
+}
 
 /**
  * @brief Maps an opaque caller session handle into its internal state structure.
  */
-P11SessionPtr_t prvSessionPointerFromHandle( CK_SESSION_HANDLE xSession )
+static P11Session_t * prvSessionPointerFromHandle( const CK_SESSION_HANDLE xSession )
 {
-    return ( P11SessionPtr_t ) xSession; /*lint !e923 Allow casting integer type to pointer for handle. */
+    /* PKCS #11 uses opaque integer handles to manage sessions.
+     * It is necessary to cast this to a pointer in order to
+     * retrieve the session. */
+    /* coverity[misra_c_2012_rule_11_4_violation] */
+    return ( P11Session_t * ) xSession;
 }
 
 /**
  * @brief Determines if an operation is in progress.
  */
-static CK_BBOOL prvOperationActive( P11SessionPtr_t pxSession )
+static CK_BBOOL prvOperationActive( const P11Session_t * pxSession )
 {
-    CK_BBOOL xResult = CK_FALSE;
+    /* coverity[misra_c_2012_rule_10_5_violation] */
```

Comment:
No explanation for the suppression - Same for all other places.

Reply:
Add a comment that points to the explanation